cli/tools/registry/mod.rs

```diff
@@ -93,157 +180,333 @@ impl PreparedPublishPackage {
   cli_options: Arc<CliOptions>,
   mapped_resolver: Arc<MappedSpecifierResolver>,
-  sloppy_imports_resolver: Option<SloppyImportsResolver>,
-  diagnostics_collector: &PublishDiagnosticsCollector,
-) -> Result<Rc<PreparedPublishPackage>, AnyError> {
-  let config_path = deno_json.specifier.to_file_path().unwrap();
-  let root_dir = config_path.parent().unwrap().to_path_buf();
-  let Some(version) = deno_json.json.version.clone() else {
-    bail!("{} is missing 'version' field", deno_json.specifier);
-  };
-  if deno_json.json.exports.is_none() {
-    let mut suggested_entrypoint = None;
+  sloppy_imports_resolver: Option<Arc<SloppyImportsResolver>>,
+}
 
-    for entrypoint in SUGGESTED_ENTRYPOINTS {
-      if root_dir.join(entrypoint).exists() {
-        suggested_entrypoint = Some(entrypoint);
-        break;
+impl PublishPreparer {
+  pub fn new(
+    graph_diagnostics_collector: GraphDiagnosticsCollector,
+    module_graph_creator: Arc<ModuleGraphCreator>,
+    source_cache: Arc<ParsedSourceCache>,
+    type_checker: Arc<TypeChecker>,
+    fs: Arc<dyn FileSystem>,
+    cli_options: Arc<CliOptions>,
+    mapped_resolver: Arc<MappedSpecifierResolver>,
+  ) -> Self {
+    let sloppy_imports_resolver = if cli_options.unstable_sloppy_imports() {
+      Some(Arc::new(SloppyImportsResolver::new(fs.clone())))
+    } else {
+      None
+    };
+    Self {
+      graph_diagnostics_collector,
+      module_graph_creator,
+      source_cache,
+      type_checker,
+      cli_options,
+      mapped_resolver,
+      sloppy_imports_resolver,
+    }
+  }
+
+  pub async fn prepare_packages_for_publishing(
+    &self,
+    allow_slow_types: bool,
+    diagnostics_collector: &PublishDiagnosticsCollector,
+    deno_json: ConfigFile,
+  ) -> Result<PreparePackagesData, AnyError> {
+    let members = deno_json.to_workspace_members()?;
+
+    if members.len() > 1 {
+      log::info!("Publishing a workspace...");
+    }
+
+    // create the module graph
+    let graph = self
+      .build_and_check_graph_for_publish(
+        allow_slow_types,
+        diagnostics_collector,
+        &members,
+      )
+      .await?;
+
+    let mut package_by_name = HashMap::with_capacity(members.len());
+    let publish_order_graph =
+      publish_order::build_publish_order_graph(&graph, &members)?;
+
+    let results = members
+      .into_iter()
+      .map(|member| {
+        let graph = graph.clone();
+        async move {
+          let package = self
+            .prepare_publish(
+              &member.package_name,
+              &member.config_file,
+              graph,
+              diagnostics_collector,
+            )
+            .await
+            .with_context(|| {
+              format!("Failed preparing '{}'.", member.package_name)
+            })?;
+          Ok::<_, AnyError>((member.package_name, package))
+        }
+        .boxed()
+      })
+      .collect::<Vec<_>>();
+    let results = deno_core::futures::future::join_all(results).await;
```

Comment:
Won't this cause problems in tests? I guess we can use `[UNORDERED_START]`...

Reply:
This was just moved from below. It seems to have been ok.